File: src/metropulse/domain/replay.py
"""Commute Replay: what a completed trip (or a month of them) actually cost
and saved, framed the way a commuter thinks about it rather than as raw
journey rows.

Every number here is a clearly-labelled estimate built from a documented
formula and a real measurement (actual ride duration, real stop
coordinates) — never a live traffic/pricing API MetroPulse doesn't have.
See :mod:`metropulse.application.intelligence.commute_impact` for the exact
constants and why each one was chosen.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime


@dataclass(frozen=True, slots=True)
class TripReplay:
    """The story of one completed trip."""

    origin_stop_id: str
    origin_name: str
    destination_stop_id: str
    destination_name: str
    started_at: datetime
    ended_at: datetime
    duration_seconds: float
    distance_km: float
    metro_fare_rupees: int
    estimated_cab_fare_rupees: int
    money_saved_rupees: int
    time_saved_seconds: float
    co2_saved_kg: float


@dataclass(frozen=True, slots=True)
class MonthlyReplay:
    """A rolling window of trips, summed — the "This Month" card."""

    period_start: datetime
    period_end: datetime
    trip_count: int
    total_distance_km: float
    total_time_saved_seconds: float
    total_money_saved_rupees: int
    total_co2_saved_kg: float

File: app/lib/domain/models/replay.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'replay.freezed.dart';
part 'replay.g.dart';

/// The story of one completed trip. Every figure here is a documented
/// estimate (see the backend's application/intelligence/commute_impact.py)
/// — never a live pricing/traffic feed — and should always read to the user
/// as an estimate, not a receipt.
@freezed
class TripReplay with _$TripReplay {
  const factory TripReplay({
    required String originStopId,
    required String originName,
    required String destinationStopId,
    required String destinationName,
    required DateTime startedAt,
    required DateTime endedAt,
    required double durationSeconds,
    required double distanceKm,
    required int metroFareRupees,
    required int estimatedCabFareRupees,
    required int moneySavedRupees,
    required double timeSavedSeconds,
    required double co2SavedKg,
  }) = _TripReplay;

  factory TripReplay.fromJson(Map<String, dynamic> json) => _$TripReplayFromJson(json);
}

/// A rolling-window summary of completed trips — the "This Month" card.
@freezed
class MonthlyReplay with _$MonthlyReplay {
  const factory MonthlyReplay({
    required DateTime periodStart,
    required DateTime periodEnd,
    required int tripCount,
    required double totalDistanceKm,
    required double totalTimeSavedSeconds,
    required int totalMoneySavedRupees,
    required double totalCo2SavedKg,
  }) = _MonthlyReplay;

  factory MonthlyReplay.fromJson(Map<String, dynamic> json) => _$MonthlyReplayFromJson(json);
}
